### PR TITLE
Add TPU and evaluation saving for GLUE finetuning

### PR DESCRIPTION
**Pull request by chenmoneygithub**

### CODE REVIEW

**Review by mattdangerw — CHANGES_REQUESTED** [review]

File: examples/bert/bert_finetune_glue.py [hunk 1]
```diff
@@ -161,12 +179,14 @@ def __init__(
             kernel_initializer=initializer,
             name="logits",
         )
+        self._drop_out = tf.keras.layers.Dropout(dropout)
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/bert/bert_finetune_glue.py]: done!

File: examples/bert/bert_finetune_glue.py [hunk 2]
```diff
@@ -177,7 +197,12 @@ def __init__(self, model_config):
         self.model_config = model_config
 
     def build(self, hp):
-        model = keras.models.load_model(FLAGS.saved_model_input, compile=False)
+        # model = keras.models.load_model(FLAGS.saved_model_input, compile=False)
+        model = keras.models.load_model(
+            "gs://chenmoney-testing-east/" + FLAGS.saved_model_input,
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/bert/bert_finetune_glue.py]: done!

File: examples/bert/bert_finetune_glue.py [hunk 3]
```diff
@@ -270,6 +306,58 @@ def preprocess_data(inputs, labels):
         f"The best hyperparameters found are:\nLearning Rate: {best_hp['lr']}"
     )
 
+    if FLAGS.saved_evaluations_output:
+        filenames = {
+            "cola": "CoLA.tsv",
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/bert/bert_finetune_glue.py:312]: yes, they require the exact name to match

File: examples/bert/bert_finetune_glue.py [hunk 4]
```diff
@@ -53,6 +53,12 @@
     "The directory to save the finetuned model.",
 )
 
+flags.DEFINE_string(
+    "saved_evaluations_output",
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/bert/bert_finetune_glue.py]: done!

File: examples/bert/bert_finetune_glue.py [hunk 5]
```diff
@@ -270,6 +306,58 @@ def preprocess_data(inputs, labels):
...
+            "qnli": ["entailment", "not_entailment"],
+            "rte": ["entailment", "not_entailment"],
+        }
+
+        filename = (
+            FLAGS.saved_evaluations_output + "/" + filenames[FLAGS.task_name]
+        )
+
+        @tf.function
+        def eval_step(iterator):
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/bert/bert_finetune_glue.py]: We need the predicted label to write to the file. `evaluate` only returns metrics?

File: examples/bert/bert_finetune_glue.py [hunk 6]
```diff
@@ -270,9 +296,57 @@ def preprocess_data(inputs, labels):
...
+            def step_fn(inputs):
+                prob = finetuning_model(inputs)
+                pred = tf.argmax(prob, -1)
+                return pred
+
+            return strategy.run(step_fn, args=(inputs,))
+
+        labelname = labelnames.get(FLAGS.task_name)
+        with tf.io.gfile.GFile(filename, "w") as f:
+            # Write the required headline for GLUE.
```
Comment (mattdangerw) [inline examples/bert/bert_finetune_glue.py:335]: I think you can use the builtin csv with a different delimiter to write tsvs
https://docs.python.org/3/library/csv.html
https://stackoverflow.com/q/29895602

We should prefer that over doing this writing outself

**Review by mattdangerw — CHANGES_REQUESTED** [review]

File: examples/bert/bert_finetune_glue.py [hunk 7]
```diff
@@ -161,12 +173,14 @@ def __init__(
             kernel_initializer=initializer,
             name="logits",
         )
+        self._drop_out_layer = tf.keras.layers.Dropout(dropout)
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/bert/bert_finetune_glue.py]: done

File: examples/bert/bert_finetune_glue.py [hunk 8]
```diff
@@ -178,6 +192,7 @@ def __init__(self, model_config):
 
     def build(self, hp):
         model = keras.models.load_model(FLAGS.saved_model_input, compile=False)
+        model = model.bert_model
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/bert/bert_finetune_glue.py]: done

File: examples/bert/bert_finetune_glue.py (hunk 6 above)
Comment (chenmoneygithub) [reply examples/bert/bert_finetune_glue.py]: done
